pkg/api_gateway/api_gateway.go

```diff
@@ -0,0 +1,202 @@
+
+	serr = buildSubstrateError(err)
+	return contractID, serr
+}
+
+func (g *apiGateway) GetNodes(farmID uint32) ([]uint32, error) {
+	log.Trace().Str("method", "GetNodes").Uint32("farm id", farmID).Msg("method called")
+	return g.sub.GetNodes(farmID)
+}
+
+func (g *apiGateway) GetPowerTarget(nodeID uint32) (power substrate.NodePower, err error) {
+	log.Trace().Str("method", "GetPowerTarget").Uint32("node id", nodeID).Msg("method called")
+	return g.sub.GetPowerTarget(nodeID)
+}
+
+func (g *apiGateway) GetTwin(id uint32) (substrate.Twin, error) {
+	log.Trace().Str("method", "GetTwin").Uint32("id", id).Msg("method called")
+	twin, err := g.sub.GetTwin(id)
+	if twin == nil {
+		twin = &substrate.Twin{}
+	}
+	return *twin, err
+}
+
+func (g *apiGateway) GetTwinByPubKey(pk []byte) (result uint32, serr pkg.SubstrateError) {
+	log.Trace().Str("method", "GetTwinByPubKey").Str("pk", hex.EncodeToString(pk)).Msg("method called")
+	twinID, err := g.sub.GetTwinByPubKey(pk)
+
+	serr = buildSubstrateError(err)
+	return twinID, serr
+}
+
+func (g *apiGateway) Report(consumptions []substrate.NruConsumption) (types.Hash, error) {
+	contractIDs := make([]uint64, 0, len(consumptions))
+	for _, v := range consumptions {
+		contractIDs = append(contractIDs, uint64(v.ContractID))
+	}
+	log.Debug().Str("method", "Report").Uints64("contract ids", contractIDs).Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.Report(g.identity, consumptions)
+}
+
+func (g *apiGateway) SetContractConsumption(resources ...substrate.ContractResources) error {
+	contractIDs := make([]uint64, 0, len(resources))
+	for _, v := range resources {
+		contractIDs = append(contractIDs, uint64(v.ContractID))
+	}
+	log.Debug().Str("method", "SetContractConsumption").Uints64("contract ids", contractIDs).Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.SetContractConsumption(g.identity, resources...)
+}
+
+func (g *apiGateway) SetNodePowerState(up bool) (hash types.Hash, err error) {
+	log.Debug().Str("method", "SetNodePowerState").Bool("up", up).Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.SetNodePowerState(g.identity, up)
+}
+
+func (g *apiGateway) UpdateNode(node substrate.Node) (uint32, error) {
+	log.Debug().Str("method", "UpdateNode").Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.UpdateNode(g.identity, node)
+}
+
+func (g *apiGateway) UpdateNodeUptimeV2(uptime uint64, timestampHint uint64) (hash types.Hash, err error) {
+	log.Debug().
+		Str("method", "UpdateNodeUptimeV2").
+		Uint64("uptime", uptime).
+		Uint64("timestamp hint", timestampHint).
+		Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.UpdateNodeUptimeV2(g.identity, uptime, timestampHint)
+}
+
+func buildSubstrateError(err error) (serr pkg.SubstrateError) {
+	if err == nil {
+		return
+	}
+
+	serr.Err = err
+	serr.Code = pkg.CodeGenericError
+
+	if errors.Is(err, substrate.ErrNotFound) {
```

Comment:
Maybe it's a good idea to also add the few remaining error codes from the substrate package  for the future and completeness of the code

![image](https://github.com/threefoldtech/zos/assets/10920323/f96b5d98-b892-4656-9baf-eeeddb23a945)